### PR TITLE
[User Model] Create pushSubscription Namespace

### PR DESCRIPTION
**Pull request by jennantilla**
# Description
## One Line Summary
Create pushSubscription namespace to handle Push Subscription methods

### Motivation
User Model requires namespacing to decouple from god OneSignal class.

### Scope
Push Subscription methods have been moved to their own namespace and will be invoked as:
```
window.plugins.OneSignal.User.pushSubscription.addObserver(function (state){
          console.log("Push Subscription state changed: " + JSON.stringify(state));
    });
```
```
window.plugins.OneSignal.User.pushSubscription.removeObserver(function (state){
          console.log("Push Subscription state changed: " + JSON.stringify(state));
    });
```
`window.plugins.OneSignal.User.pushSubscription.optIn();`
`window.plugins.OneSignal.User.pushSubscription.optOut();`

Separate methods created in order to access pushSubscription properties with the following API:
`let pushId = window.plugins.OneSignal.User.pushSubscription.id;`
`let token = window.plugins.OneSignal.User.pushSubscription.token;`
`let optedIn = window.plugins.OneSignal.User.pushSubscription.optedIn;`

### OPTIONAL - Other
This PR only encompasses changes to the main plugin interface and iOS. Separate Android PR to follow.

# Testing
## Unit testing
No unit testing was used.

## Manual testing
**RECOMMEND - OPTIONAL -** Explain what scenarios were tested and the environment.
Successfully tested methods on iPad running iOS 15.2

# Affected code checklist
   - [ ] Notifications
      - [ ] Display
      - [ ] Open
      - [ ] Push Processing
      - [ ] Confirm Deliveries
   - [ ] Outcomes
   - [ ] Sessions
   - [ ] In-App Messaging
   - [ ] REST API requests
   - [X] Public API changes

# Checklist
## Overview
   - [X] I have filled out all **REQUIRED** sections above
   - [X] PR does one thing
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR
   - [X] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing
   - [X] I have included test coverage for these changes, or explained why they are not needed
   - [X] All automated tests pass, or I explained why that is not possible
   - [X] I have personally tested this on my device, or explained why that is not possible

## Final pass
   - [X] Code is as readable as possible.
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
   - [X] I have reviewed this PR myself, ensuring it meets each checklist item
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.

- - -
This change is [Reviewable](https://reviewable.io/reviews/OneSignal/OneSignal-Cordova-SDK/852)

### CODE REVIEW

**Comment by jennantilla** [comment]
@jkasten2 - this is the solution @nan-li and I came up with for accessing pushSubscription properties. This way devs can use `let token = OneSignal.User.pushSubscription.token` 
Please let me know if any concerns!

File: src/ios/OneSignalPush.m [hunk 1]
```diff
@@ -275,11 +275,51 @@ - (void)addPermissionObserver:(CDVInvokedUrlCommand*)command {
...
+    }
+}
+
+- (void)getToken:(CDVInvokedUrlCommand*)command {
+    NSString *token = OneSignal.User.pushSubscription.token;
+    if (token) {
+        NSDictionary *result = @{
+            @"value" : token
+    };
+    successCallback(command.callbackId, result);
```
Comment (brismithers) [inline src/ios/OneSignalPush.m:305]: nit to fix indentation on this line and previous line

**Review by jkasten2 — APPROVED** [review]
:lgtm:

> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/OneSignal/OneSignal-Cordova-SDK/852#-:-NNNe8Gb2BzlySAOafyS:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @emawby, @fhboswell, and @nan-li)

**Review by nan-li — COMMENTED** [review]
> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/OneSignal/OneSignal-Cordova-SDK/852)* status: 4 of 5 files reviewed, 3 unresolved discussions

___
*[`src/ios/OneSignalPush.m` line 305 at r1](https://reviewable.io/reviews/OneSignal/OneSignal-Cordova-SDK/852#-NNNgOZeD6vXp0wMXSB-:-NNNgOZeD6vXp0wMXSB0:b-yul2qc) ([raw file](https://github.com/OneSignal/OneSignal-Cordova-SDK/blob/f686c6138fccb46587735af2f746ed428ab26754/src/ios/OneSignalPush.m#L305)):*
> ```objc
>             @"value" : token
>     };
>     successCallback(command.callbackId, result);
> ```

nit, indent so its clear these are happening within `if token` and not outside.



_Code quote:_
```objc
    };
    successCallback(command.callbackId, result);
```
___
*[`www/PushSubscriptionNamespace.ts` line 23 at r1](https://reviewable.io/reviews/OneSignal/OneSignal-Cordova-SDK/852#-NNNiJzT2zxQcxU7bTPZ:-NNNiJzT2zxQcxU7bTP_:b-1ochm6) ([raw file](https://github.com/OneSignal/OneSignal-Cordova-SDK/blob/f686c6138fccb46587735af2f746ed428ab26754/www/PushSubscriptionNamespace.ts#L23)):*
> ```XML
>      * Sets initial Push Subscription properties and adds observer for changes
>      */
>     setPropertiesAndObserver():void {
> ```

Since this is for internal use, ideally, it won't show up for the app developer when the call `OneSignal.User.pushSubscription...` and think it's a public method they can use. But I think this does. Is there any way to hide this but still be able to call it from `index.ts`? Otherwise, for now we can maybe change this method to be named `__setPropertiesAndObserver()` so at least if it shows up, people are less likely to call it.
___
*[`www/PushSubscriptionNamespace.ts` line 89 at r1](https://reviewable.io/reviews/OneSignal/OneSignal-Cordova-SDK/852#-NNNjYOn92rUFbC7p9jP:-NNNjYOn92rUFbC7p9jQ:b-554ytc) ([raw file](https://github.com/OneSignal/OneSignal-Cordova-SDK/blob/f686c6138fccb46587735af2f746ed428ab26754/www/PushSubscriptionNamespace.ts#L89)):*
> ```XML
>      */
>     removeObserver(): void {
>         window.cordova.exec(function(){}, function(){}, "OneSignalPush", "removePushSubscriptionObserver", []);
> ```

This won't quite work as we have it. If the app developer removes their observer, we won't be able to observe the push subscription anymore either as we will stop getting callbacks from the native SDK. I think it's sufficient to just remove this person's subscription observer from the `_subscriptionObserverList` so they won't get triggered but we still will.

**Comment by nan-li** [comment]
Sorry @jennantilla reviewed too late. We can look at the last 2 items when reviewing the permission observer since that will have similar issue. The nit is a non-blocker.

**Pull request merged**